jetty-server/src/main/java/org/eclipse/jetty/server/handler/HandlerCollection.java

```diff
@@ -168,16 +174,22 @@ public void addHandler(Handler handler)
      */
     public void prependHandler(Handler handler)
     {
-        setHandlers(ArrayUtil.prependToArray(handler, getHandlers(), Handler.class));
+        synchronized (this)
+        {
+            setHandlers(ArrayUtil.prependToArray(handler, getHandlers(), Handler.class));
+        }
     }
 
     /* ------------------------------------------------------------ */
     public void removeHandler(Handler handler)
     {
-        Handler[] handlers = getHandlers();
+        synchronized (this)
+        {
+            Handler[] handlers = getHandlers();
```

Comment:
If another thread uses getHandlers() while this is occuring, wouldn't this be a concurrent modification exception?

Reply:
no, because the caller of getHandler would see a version of handlers either just before or just after this removeHandler  sets the handler array with a new array.   So we have atomic change and the array they look at will never change.     Now the handlers in that array may also change state if they are removed, but we have to rely on them also being threadsafe with state changes.